File: BucketSort.java
import java.io.*; 
import java.util.*;

public class bucketSort {

	int minData; 
	int maxData;
	int offset; 
	int[] bucketAry; 
	
	bucketSort(){
		minData = 0; 
		maxData = 0; 
		offset = 0; 
	}
	void findMinMax(int data){
		if(data > maxData){
			maxData = data; 
		}
		if(data < minData){
			minData = data; 
		}
	}
	int getIndex(int data){
		return data - offset; 
	}
	void printBucketAry(BufferedWriter outFile2){
		try {
			outFile2.write("Index: ");
			for(int i = 0; i <=18; i++){
				if(i < 10){
					outFile2.write("0"+i+" "); 
				}
				else
					outFile2.write(i + " ");
			}
			outFile2.newLine();
			
			outFile2.write("Value:  ");
			for(int i = 0; i <=18; i++){
				outFile2.write(bucketAry[i]+"  "); 
			}
			outFile2.newLine();
			outFile2.close(); 
			
		} catch (IOException e) {
			e.printStackTrace();
		}
		
	}
	void printSortedData(BufferedWriter outFile1) {
		try {
			int kVal = 1;
			outFile1.write("Sorted Values using BucketSort Implementation:");
			outFile1.newLine();
			for (int i = 0; i < bucketAry.length; i++) {
				while (bucketAry[i] != 0) {
					int val = i + offset; 
					outFile1.write("Output #" + kVal + ": " + val);
					outFile1.newLine();

					bucketAry[i]--;
					kVal++;
				}
			}
		} catch (IOException e) {
			e.printStackTrace();
		}
	}

	public static void main(String[] args){

		try {
			bucketSort SortedAry = new bucketSort(); 
			int num = 0; 
			Scanner inFile = new Scanner(new FileReader(args[0]));
			num= inFile.nextInt();
			SortedAry.minData = num; 
			SortedAry.maxData = num; 
			while(inFile.hasNextInt()){
				num = inFile.nextInt(); 
				SortedAry.findMinMax(num);
			}
			inFile.close(); 
			
			int minSize = SortedAry.maxData - SortedAry.minData +1; 
			SortedAry.bucketAry = new int[minSize];

			SortedAry.offset = SortedAry.minData;
			
			inFile = new Scanner(new FileReader(args[0]));
			
			File outPath1 = new File(args[2]);
			File outPath2 = new File(args[1]);
			BufferedWriter outFile1 = new BufferedWriter(new FileWriter(outPath1));
			BufferedWriter outFile2 = new BufferedWriter(new FileWriter(outPath2));
			outFile2.write("Updating values in bucketSort array...");
			outFile2.newLine(); 
			while(inFile.hasNextInt()){
				num = inFile.nextInt(); 
				int index = SortedAry.getIndex(num); 
				SortedAry.bucketAry[index]++; 
				outFile2.write("BucketAry[" + index + "] = " + SortedAry.bucketAry[index]);
				outFile2.newLine();  
			}
			outFile2.write("Values updated.");
			outFile2.newLine(); 
			outFile2.newLine();
			inFile.close(); 
			
			SortedAry.printBucketAry(outFile2);
			SortedAry.printSortedData(outFile1);
			
			outFile1.close();
			outFile2.close(); 
		} catch (FileNotFoundException e) {
			e.printStackTrace();
		} catch (IOException e) {
			e.printStackTrace();
		}
	
	}
}

File: BucketSort.cpp
//March 7, 2016 
//Algorithms 
//Implementation of bucket sort using the minimum # of buckets

#include <stdio.h>
#include <stdlib.h>
#include <iostream>
#include <fstream>

using namespace std; 

class bucketSort {
private:
    int minData; 
    int maxData; 
    int bucketSize; 
    int offset; 
    int* bucketAry;
public: 
    bucketSort(); 
    ~bucketSort(); 
    void findMinMax(int data); 
    int getIndex(int value); 
    void printBucketArray(ofstream &outFile2);
    void printSortedData(ofstream &outFile1);
    //Getters and setters 
    int getMin();
    int getMax();
    void setMin(int initMin); 
    void setMax(int initMax); 
    int* getBucket();
    void setBucketSize(int size);
    void setOffset(int val);   
};

bucketSort::bucketSort(){}
bucketSort::~bucketSort(){}
void bucketSort::findMinMax(int data){
    if(data > maxData){
        maxData = data; 
    }
    if(data < minData){
        minData = data;
    } 
}
int bucketSort::getIndex(int value){
    return value - offset;   
}
void bucketSort::setBucketSize(int size){
    bucketAry = new int[size+1]; 
    //initialize all values to 0
    for(int i = 0; i<=size; i++){
        bucketAry[i] = 0;  
    }
}
void bucketSort::printBucketArray(ofstream &outFile2){
    outFile2 << "Index: "; 
    for(int i =0; i < 19; i++){
        if(i < 10){
            outFile2 << "0" << i << " ";
        }
        else{
            outFile2 << i << " " ; 
        }    
    }
    outFile2 << endl; 
    outFile2 << "Value:  "; 
    for(int i =0; i < 19; i++){
        outFile2 << bucketAry[i] << "  "; 
    }
    outFile2 << endl; 
}
void bucketSort::printSortedData(ofstream &outFile1){
    int length = maxData - minData +1;
    int kVal = 1;
    outFile1 << "Sorted list using bucketSort Implementation: " << endl; 
    for(int i =0; i < length; i++){
        while(bucketAry[i]!=0){
            outFile1 << "Output #" << kVal << ": " << i + offset << endl; 
            kVal++;
            bucketAry[i]--; 
        }
    }
}
int bucketSort::getMin(){return minData;}
int bucketSort::getMax(){return maxData;}
void bucketSort::setMin(int initMin){minData = initMin;}
void bucketSort::setMax(int initMax){maxData = initMax;}
void bucketSort::setOffset(int val){offset = val;}
int* bucketSort::getBucket(){return bucketAry;}


int main(int argc, char** argv) {
    
    int data; 
    
    string filePath = argv[1];
    ifstream inFile(filePath);
    ofstream outFile1(argv[2]);
    ofstream outFile2(argv[3]);
 
    bucketSort bucketSorted;
    
    //go through infile and find min/max data
    inFile >> data; 
    bucketSorted.setMin(data); 
    bucketSorted.setMax(data);
    while(inFile >> data){
       // inFile >> data; 
        bucketSorted.findMinMax(data);
    }
    inFile.close(); 
    
    //Compute the minimum bucket size 
    int bucketSize = bucketSorted.getMax() - bucketSorted.getMin();
    
    //Compute the offset
    bucketSorted.setOffset(bucketSorted.getMin()); 
    
    //Dynamically allocate the bucketAry array to bucketSize
    bucketSorted.setBucketSize(bucketSize);
    
    //re-open the input file
    inFile.open(filePath); 
    outFile2 << "Updating values in array with bucketSort implementation..." << endl; 
    while(inFile >> data){
       // inFile >> data; 
        
        //Compute the index
        int index = bucketSorted.getIndex(data);
        
        //Update the array 
        bucketSorted.getBucket()[index]++;
        
        outFile2 << "BucketAry[" << index <<"]= " << bucketSorted.getBucket()[index] << endl;
    }
    outFile2 << endl << "Updating array complete." << endl << endl; 
    
    //Call printBucketArray
    bucketSorted.printBucketArray(outFile2);
    
    //Call printSortedData 
    bucketSorted.printSortedData(outFile1); 

    inFile.close();
    outFile1.close(); 
    outFile2.close(); 
}